### ISSUE TITLE
[Java] Hiding of parent static field is not flagged as violation

### ISSUE THREAD

**Issue by Fernal73** [pmd/pmd#1987]
**Affects PMD Version:** 
6.17.0

**Rule:**
No rule as yet. 

**Description:**
The following hiding of a static field in parent class is not flagged as a violation by any of the existing PMD rules. A new rule is needed. 

**Code Sample demonstrating the issue:**

```java
package javapuzzles;

class Jeopardy {
  public static final String PRIZE = "$64,000";
}

public class DoubleJeopardy extends Jeopardy {
  public static final String PRIZE = "2 cents";

  public static void main(String[] args) {
    System.out.println(DoubleJeopardy.PRIZE);
  }
}
```

**Running PMD through:** *[CLI | Ant ]*

**Comment by UncleOwen** [comment]
I don't see the problem here.

```java
class Jeopardy {
  public static final String PRIZE = "$64,000";
}

public class DoubleJeopardy extends Jeopardy {
  public static final String PRIZE = "2 cents";

  public static void main(String[] args) {
    System.out.println(Jeopardy.PRIZE);        // prints "$64,000"
    System.out.println(DoubleJeopardy.PRIZE);  // prints "2 cents"
  }
}
```

Where is the unexpected, tricky or otherwise errorprone behavior?

If anything, I'd argue we'd need a rule that triggers when a static member is accessed via a subclass or variable. (Do we have a rule like that already? - no, but there's #2933)

**Comment by adangel** [comment]
If you are not familiar with how Java implemented static fields (no inheritance!!), you might be thinking, that you could overwrite the `PRIZE` field on DoubleJeopardy - which you can't (Jeopardy.PRIZE is unaffected). To avoid such confusion, a project might decide to forbid declaring a static field in a subclass, that hides a field from a superclass.
Also, if you access the field unqualified, it might be confusing which one is being used (especially if you move code between the two classes). E.g. using `PRICE` in DoubleJeopardy hides the field `PRICE` in Jeopardy.

The OP code sample probably should have been:
...
```java
    System.out.println(PRICE); // which PRICE: from DoubleJeopardy or Jeopardy?
```

I think, it would be useful to have such a rule.

> we'd need a rule that triggers when a static member is accessed via a subclass or variable. (Do we have a rule like that already?)

That's also correct but would probably be a different rule. It's another confusing part of how Java did it: For static fields, it doesn't depend on the runtime type, but on the static compile time type of the variable... We could extend it to catch unqualified access as well.
And no, I don't know, if we have already such a rule.

**Comment by linusjf** [comment]
The code example is from [Java Puzzlers
Traps, Pitfalls, and Corner Cases By Joshua Bloch and Neal Gafter](http://www.javapuzzlers.com/). Puzzle 72: Final Jeopardy. 


The authors explain it as below:

```
Because the PRIZE field in Jeopardy is declared public and final, you might
think that the language would prevent you from reusing this field name in a sub-
class. After all, final methods cannot be overridden or hidden. If you tried the pro-
gram, you found that it compiles without a hitch and prints 2 cents. What went
wrong?
It turns out that the final modifier means something completely different
on methods and fields. On a method, final means that the method may not be
overridden (for instance methods) or hidden (for static methods) [JLS 8.4.3.3].
 On
a field, final means the field may not be assigned more than once [JLS 8.3.1.2].

The keyword is the same, but the behavior is unrelated.

In the program, the final field DoubleJeopardy.PRIZE hides final field
Jeopardy.PRIZE, for a net loss of $63,999.98. Although it is possible to hide
fields, it is generally a bad idea. As we discussed in Puzzle 66, hiding fields can
violate subsumption and confound our intuition about the relationship between
types and their members.

If you want to guarantee the prize in the Jeopardy class even while preserving
the ability to subclass it, use a final method instead of a final field:

class Jeopardy {
private static final String PRIZE = "$64,000";
public static final String prize() {
return PRIZE;
}
}

For language designers, the lesson is to avoid reusing the same keyword for
unrelated concepts. A keyword should be reused only for closely related concepts,
where it helps programmers build an intuition about the relationship among the
language features in question. In the case of the Java’s final keyword, reuse leads
to confusion. It should be noted that as a language ages, there is a natural tendency
to reuse keywords for unrelated concepts. This avoids the need to introduce new
keywords, which is enormously destabilizing. When language designers do this,
they are generally choosing the lesser of two evils.
In summary, avoid reusing names for unrelated variables or unrelated con-
cepts. Using distinct names for unrelated concepts helps readers and programmers
to keep the concepts separate.
```

**Comment by UncleOwen** [comment]
@adangel: Yes, that example is better, thanks.

@linusjf: I've read that book! Note, that in the book, the question is always: "Does the code compile - and if so, what output does it produce?" The author's explanation focuses on the first part.

**Comment by UncleOwen** [comment]
If we want a rule like this - should this only apply to static variable? If hiding static variables is a problem, surely hiding non-static ones is a problem as well?

**Comment by Asobu01** [comment]
Happy to claim pmd/pmd#1987.
I read "[Java] Hiding of parent static field is not flagged as violation" and my first move is reproducing the reported issue first, then applying a focused fix with clear validation.
ETA: first reviewable PR in about 8h.
I can post a reproducible checklist/log snippet in ~30 minutes so you can verify execution speed before the full PR.
I will include repro + root-cause summary, focused patch (plus tests/docs if needed), and review-ready PR with validation notes.
Thanks. @Asobu01

**Comment by UncleOwen** [comment]
@Asobu01 Sure go ahead. Note that we're all volunteers here. No need for ETAs, especially ETAs in minutes or hours.

**Comment by Asobu01** [comment]
Quick follow-up on `pmd/pmd#1987`. I am still ready to start immediately and can post the first repro/checklist note before the draft PR if helpful. If you want me on this one, I can begin now. (@Asobu01)
